hw06-spring-orm-jpa

IlyaLeshin opened the pull request:
ДЗ №6: Переписать приложение для хранения книг на ORM

IlyaLeshin committed:
hw06-spring-orm-jpa (ДЗ №6: Переписать приложение для хранения книг на ORM)

created template

IlyaLeshin committed:
added comments + ORM

IlyaLeshin committed:
added comments into Book.java

IlyaLeshin committed:
deleted entity-graph, changed entity fields to LAZY, added Transaction(readOnly=true) on getMethods into BookServiceImpl

IlyaLeshin committed:
added/changed tests

IlyaLeshin committed:
added JpaCommentRepositoryTest.java

IlyaLeshin committed:
added and fixed tests

IlyaLeshin committed:
deleted Transactional(readOnly = true) from findAll() GenreServiceImpl

kataus commented on hw06-spring-orm-jpa/src/main/java/ru/otus/hw/models/Book.java:
```diff
@@ -0,0 +1,59 @@
+    private Author author;
+
+    @Fetch(FetchMode.SUBSELECT)
+    @ManyToMany(fetch = FetchType.LAZY)
```
Нет смысла повторять дефолтные настройки, это только лишний код

IlyaLeshin replied:
убрал дефолтные настройки

kataus commented on hw06-spring-orm-jpa/src/main/java/ru/otus/hw/models/Comment.java:
```diff
@@ -0,0 +1,36 @@
+    @Column(name = "text", nullable = false)
+    private String text;
+
+    @ManyToOne(fetch = FetchType.LAZY)
```
А вот тут про LAZY я бы сильно посмотрил. Комментарий без книги не имеет функционального смысла

IlyaLeshin replied:
согласен, убрал Lazy, оставил дефолтное значение

kataus commented on hw06-spring-orm-jpa/src/main/java/ru/otus/hw/repositories/JpaAuthorRepository.java:
```diff
@@ -0,0 +1,29 @@
+import java.util.List;
+import java.util.Optional;
+
+@Repository
```
Что тут используется от аннотации @Repository? Она несколько больше чем @Component или @Service. Если она появляется - то мы ожидаем использование её особенностей

IlyaLeshin replied:
заменил на @Component

kataus commented on hw06-spring-orm-jpa/src/main/java/ru/otus/hw/repositories/JpaCommentRepository.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @Override
+    public List<Comment> findAllByBookId(long bookId) {
```
Давайте попробуем работать со связями через ОРМ. Этот метод удалим и решим задачу в бизнес-слое. При этом связь book->comments должна быть ленивой. Без EAGER и подругзки через графы. Задача - именно ленивую связь разрешить

IlyaLeshin replied:
Удалил метод в репозитории, переделал метод в сервисе

IlyaLeshin committed:
fixed after review

deleted default FetchType in models,
deleted method findAllByBookId in CommentRepository and recode CommentServiceImpl,
changed Repository annotations,
fixed tests

IlyaLeshin committed:
Merge branch 'main' into hw06-spring-orm-jpa
